cloudimg/ms_azure.py

```diff
@@ -0,0 +1,513 @@
+        if blob_client.exists():
+            log.info(
+                "The image \"%s\" already exists in container \"%s\"",
+                object_name,
+                container_name,
+            )
+            return
+
+        log.info('Uploading %s to container %s', image_path, container_name)
+        log.info('Uploading %s with name %s', image_path, object_name)
+        if image_path.startswith("http:") or image_path.startswith("https:"):
+            blob_client.upload_blob_from_url(source_url=image_path, **kwargs)
+        else:
+            with open(image_path, "rb") as data:
+                blob_client.upload_blob(data=data, **kwargs)
+        log.info(self._upload_progress)
+        log.info('Successfully uploaded %s', image_path)
+        return blob_client
+
+    def publish(self, metadata):
+        """
+        Take some metadata about a VHD image and upload it to Azure.
+
+        This method relies on unique image/metadata names in order to do the
+        least amount of work possible.
+
+        Args:
+            metadata (AzurePublishingMetadata): Metadata about the VHD image
+
+        Returns:
+            The BlobProperties with the data from expected blob.
+        """
+        log.info(
+            'Searching for image: %s in container %s',
+            metadata.image_name,
+            metadata.container,
+        )
+        blob = self.get_object_by_name(
+                    container=metadata.container,
+                    name=metadata.image_name,
+                )
+
+        if not blob:
+            log.info('Image does not exist: %s', metadata.image_name)
+            log.info('Searching for tags: %s', metadata.tags)
+
+            filtered_blob = self.filter_object_by_tags(metadata.tags)
+
+            if not filtered_blob:
+                log.error("Image not found with tags \"%s\"", metadata.tags)
+                blob = self.upload_to_container(
+                    image_path=metadata.image_path,
+                    container_name=metadata.container,
+                    object_name=metadata.object_name,
+                    tags=metadata.tags,
+                )
+            else:
+                log.info(
+                    "Image already exists on container \"%s\""
+                    " with name \"%s\"",
+                    filtered_blob.container_name,
+                    filtered_blob.name,
+                )
+
+                blob = self.get_object_by_name(
+                    container=filtered_blob.container_name,
+                    name=filtered_blob.name,
+                )
+
+        else:
+            log.info("Image already exists with name \"%s\"", blob.blob_name)
+
+        log.info('Image published: %s/%s', blob.container_name, blob.blob_name)
+        return blob.get_blob_properties()
+
+    def get_blob_sas_uri(self, blob_props):
+        """
+        Get a SAS URL of a blob using its Storage account's connection string
```

Comment:
Pretty sure that a single newline is translated to a space everywhere doc strings are rendered, so this would probably be rendered as:

"Get a SAS URL of a blob using its Storage account's connection string The SAS URL is with ..."

So it ought to have punctuation at least.  I think the ideal format is usually one sentence, then a blank line, then more details in later paragraphs like:

```suggestion
        Get a SAS URL of a blob using its Storage account's connection string.

```

Reply:
Thanks!

Fixed now